src/service/src/transport.rs

```diff
@@ -0,0 +1,392 @@
+
+/// Serve a single CTP connection.
+async fn serve_connection<In, Out, H>(
+    stream: Stream,
+    mut handler: H,
+    version: Version,
+) -> anyhow::Result<Infallible>
+where
+    In: Message,
+    Out: Message,
+    H: GenericClient<In, Out>,
+{
+    let mut conn = Connection::start(stream, version).await?;
+
+    loop {
+        tokio::select! {
+            // `Connection::recv` is documented to be cancel safe.
+            inbound = conn.recv() => {
+                let msg = inbound?;
+                handler.send(msg).await?;
+            },
+            // `GenericClient::recv` is documented to be cancel safe.
+            outbound = handler.recv() => match outbound? {
+                Some(msg) => conn.send(msg).await?,
+                None => bail!("client disconnected"),
+            },
+        }
+    }
+}
+
+/// An active CTP connection.
+///
+/// This type encapsulates the core connection logic. It is used by both the client and the server
+/// implementation, with swapped `Out`/`In` types.
+///
+/// Each connection spawns two tasks:
+///
+///  * The send task is responsible for encoding and sending enqueued messages.
+///  * The recv task is responsible for receiving and decoding messages from the peer.
+///
+/// The separation into tasks provides some performance isolation between the sending and the
+/// receiving half of the connection.
+#[derive(Debug)]
+struct Connection<Out, In> {
+    /// Message sender connected to the send task.
+    msg_tx: mpsc::Sender<Out>,
+    /// Message receiver connected to the receive task.
+    msg_rx: mpsc::Receiver<In>,
+    /// Receiver for errors encountered by connection tasks.
+    error_rx: watch::Receiver<String>,
+
+    /// Handles to connection tasks.
+    _tasks: [AbortOnDropHandle<()>; 2],
+}
+
+impl<Out: Message, In: Message> Connection<Out, In> {
+    /// Start a new connection wrapping the given stream.
+    async fn start(stream: Stream, version: Version) -> anyhow::Result<Self> {
+        let (mut reader, mut writer) = stream.split();
+
+        handshake(&mut reader, &mut writer, version).await?;
+
+        let (out_tx, out_rx) = mpsc::channel(1024);
+        let (in_tx, in_rx) = mpsc::channel(1024);
+        // Initialize the error channel with a default error to return if none of the tasks
+        // produced an error.
+        let (error_tx, error_rx) = watch::channel("connection closed".into());
+
+        let send_task = mz_ore::task::spawn(
+            || "ctp::send",
+            Self::run_send_task(writer, out_rx, error_tx.clone()),
+        );
+        let recv_task =
+            mz_ore::task::spawn(|| "ctp::recv", Self::run_recv_task(reader, in_tx, error_tx));
+
+        Ok(Self {
+            msg_tx: out_tx,
+            msg_rx: in_rx,
+            error_rx,
+            _tasks: [send_task.abort_on_drop(), recv_task.abort_on_drop()],
+        })
+    }
+
+    /// Enqueue a message for sending.
+    async fn send(&mut self, msg: Out) -> anyhow::Result<()> {
```

Comment:
What about cancellation safety on the send path? The channel is bounded I think, so depending on how we call this function, we could block here.